Add rollover API to switch index aliases given some predicates

areek opened the pull request:
The rollover index API allows to switch the index pointed to by an alias given some predicates.
In order to rollover an index, the provided alias has to point to a single index. Upon satisfying
any of the predicates, the alias is switched to point to the rollover index, if the rollover index
already exists, the rollover fails.

This API is syntactic sugar for changing the index pointed to by an alias given some predicate.

Suppose there is an index named `index1` that has an alias `index_alias`.
To rollover `index_alias` to point to a new index:

```bash
$ curl -XPOST 'http://localhost:9200/index_alias/_rollover/index2' -d '{
    "conditions" : {
        "max_age": "7d", <1>
        "max_docs": 1000 <2>
    }
}'
```
<1> Sets a condition that the index has to be at least 7 days old
<2> Sets a condition that the index has to have at least a 1000 documents

The API call above switches the index pointed to by `index_alias` from `index1` to `index2`, if any
of the conditions are met. `index2` is created (using matching indices-templates if available).
The API call returns immediately if none of the conditions are met.

The rollover API can be used without specifying the name for the new index. In this case, the API requires
the old concrete index name to have `{index_prefix}-{num}` format, as rollover index name is generated
following `{index_prefix}-{num+1}` format.

The `_rollover` API is similar to `indices-create-index` and accepts `settings`, `mappings` and `aliases` to override the index create request for the rollover index.

```bash
$ curl -XPOST 'http://localhost:9200/index_alias/_rollover' -d '{
    "conditions" : {
        "max_age": "7d",
        "max_docs": 1000
    },
    "settings": { <1>
        "index.number_of_shards": 2
    }
}'
```
<1> Set settings to override matching index template, `mappings` and `aliases` can also be provided.

The rollover API supports `dry_run` mode, where request conditions can be checked without performing the actual rollover. The `dry_run` mode can be enabled by setting `dry_run=true` as a request parameter.

An example response for the index rollover API:

```
{
	"old_index": "index-1", <1>
	"new_index": "index-2", <2>
	"rolled_over": true, <3>
	"dry_run": false, <4>
	"conditions": { <5>
	    "[max_age: 7d]": true,
	    "[max_docs: 1000]": true
	}
}
```

<1> name of the index the alias was pointing to
<2> name of the index the alias currently points to
<3> whether the alias switch was successful
<4> whether the rollover was dry run
<5> status of the evaluated request conditions

closes #18647

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:
```diff
@@ -0,0 +1,164 @@
+        super.readFrom(in);
+        sourceAlias = in.readString();
+        optionalTargetAlias = in.readOptionalString();
+        int size = in.readVInt();
```
I think the "more modern" way to do this is to make Condition implement Writeable and use readList and writeList for this.

s1monw replied:
yeah I think it should be NameWriteable though

s1monw replied:
and have a subclass per condition?

nik9000 replied:
I think NamedWriteable gives you the flexibility to do more with conditions in a backwards compatible way later. So, yeah, I think NamedWriteable is better than just Writeable.

areek replied:
Now condition implements NamedWritable and has a subclass per condition

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    public void source(BytesReference source) {
+        XContentType xContentType = XContentFactory.xContentType(source);
```
I think we have ObjectParser for this kind of thing?

s1monw replied:
++

nik9000 replied:
I think one thing that ObjectParser would do for this is that it'd make it really simple if we had to add conditions that require more than one value to configure them.

areek replied:
Now rollover source parsing is done through ObjectParser

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -0,0 +1,257 @@
+                            @Override
+                            public void onFailure(Throwable t) {
+                                if (t instanceof IndexAlreadyExistsException) {
+                                    logger.trace("[{}] failed to create rollover index", t, updateRequest.index());
```
In this case do you want to just move on rather than fail?

s1monw replied:
I think we should fail in such a condition and just pass on the throwable

nik9000 replied:
If so then we don't need the if statement at all, right?

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -0,0 +1,257 @@
+                                       long creationDate) {
+        for (Condition condition: requestConditions) {
+            switch (condition.getType()) {
+                case MAX_SIZE:
```
This feels like it maybe should be an abstract method in ConditionType and each type overrides?

nik9000 replied:
Or, like @s1monw says, a subclass of Condition per condition type.

nik9000 commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.rollover/10_basic.yaml:
```diff
@@ -0,0 +1,69 @@
+  - match: { _index:   logs-1 }
+  - match: { _type:    test   }
+  - match: { _id:      "2"     }
+  - match: { _source:  { foo: "hello world" } }
```
It'd be cool to do a search through the search alias I think.

nik9000 commented on core/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java:
```diff
@@ -0,0 +1,72 @@
+    @Inject
+    public RestRolloverIndexAction(Settings settings, RestController controller, Client client) {
+        super(settings, client);
+        controller.registerHandler(RestRequest.Method.PUT, "/{alias}/_rollover/{optional_alias}", this);
```
PUT feels funny here.  Maybe just POST? It feels like you are taking a rollover action not creating a rollover entity.

s1monw replied:
I think we have PUT everywhere else too but @clintongormley will know more

s1monw replied:
maybe we should also have a GET version that only checks if the conditions are met? ie some sort of simluate mode?

nik9000 replied:
I think GET'd be cool if we returned the matching conditions.

clintongormley replied:
> I think we have PUT everywhere else too but @clintongormley will know more

No, PUT and POST are not always duplicates of each other.  I agree with @nik9000, POST feels like the right choice here.

areek replied:
I have removed the PUT method and added a GET method, which will simulate the rollover.

nik9000 commented on core/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java:
```diff
@@ -0,0 +1,72 @@
+        }
+        rolloverIndexRequest.timeout(request.paramAsTime("timeout", rolloverIndexRequest.timeout()));
+        rolloverIndexRequest.masterNodeTimeout(request.paramAsTime("master_timeout", rolloverIndexRequest.masterNodeTimeout()));
+        client.admin().indices().rolloverIndex(rolloverIndexRequest, new RestBuilderListener<RolloverResponse>(channel) {
```
Maybe implement ToXContent on the response and use RestToXContentListener?

s1monw replied:
++

areek replied:
done

nik9000 commented:
cool!

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/Condition.java:
```diff
@@ -0,0 +1,105 @@
+
+import java.util.Locale;
+
+public class Condition {
```
I think Condition should be a `NamedWriteable` check for `NamedWriteableRegistry` how it's used

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:
```diff
@@ -0,0 +1,164 @@
+
+    private String sourceAlias;
+    private String optionalTargetAlias;
+    private List<Condition> conditions = new ArrayList<>();
```
should this be a set of conditions?

areek replied:
yes, changed

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:
```diff
@@ -0,0 +1,164 @@
+        this.optionalTargetAlias = optionalTargetAlias;
+    }
+
+    public void addMaxIndexAgeCondition(String age) {
```
I think those methods should take int instead

nik9000 replied:
max age is a TimeValue though?

areek replied:
Changed to take accept TimeValue

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:
```diff
@@ -0,0 +1,164 @@
+
+    RolloverRequest() {}
+
+    public RolloverRequest(String sourceAlias, String optionalTargetAlias) {
```
what is source vs. target? I think we have a write alias and search alias and an index name?

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverResponse.java:
```diff
@@ -0,0 +1,62 @@
+public final class RolloverResponse extends ActionResponse {
+
+    private String oldIndex;
+    private String newIndex;
```
I think we should also return which condition met? a string of some sort?

s1monw replied:
also should we return if we rolled over the index at all? I would expect a boolean as well?

nik9000 replied:
I think both the indexes are the same when it wasn't changed. As in "I didn't make a new index, keep using the old one". I might prefer newIndex be null if we didn't roll over?

I think it'd be nice to have a list/set of the conditions that matched. Then they'd have to implement ToXContent but that is fine.

areek replied:
I have enhanced the response to include what conditions were met, whether the rollover was simulated and if the rollover index had to be created automatically.

s1monw commented:
I like the overall approach! Yet, I think we should work more on:

 * the request, I think there is too much magic involved and we should really just ask for the target index name in the request
 * on the response end I think we should have a boolean that says if we rolled over or not. if we did we can also return which conditions met and which index was created etc.
 * I think we need a simulate flag to check only the conditions but don't roll over?
 * After mulling on it I think we should only have age and num docs as conditions, index size is arbitrary and might be completely off due to pending merges etc.

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    public void addMaxIndexSizeCondition(String size) {
+        addCondition(Condition.ConditionType.MAX_SIZE, Condition.ConditionType.parseFromString(Condition.ConditionType.MAX_SIZE, size));
```
And this is a SizeValue?

areek replied:
nuked the Index size condition as per simon's comment

nik9000 commented on core/src/test/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverActionTests.java:
```diff
@@ -0,0 +1,192 @@
+        }
+
+        try {
+            TransportRolloverAction.validate(metaData, new RolloverRequest(randomFrom(index1, index2),
```
Probably should use `expectThrows`. Is marginally easier to read and less typing to maintain.

areek replied:
changed

nik9000 commented on core/src/test/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverActionTests.java:
```diff
@@ -0,0 +1,192 @@
+    public void testGenerateRolloverIndexName() throws Exception {
+        String index = randomAsciiOfLength(10);
+        assertThat(TransportRolloverAction.generateRolloverIndexName(index), not(equalTo(index)));
+        assertThat(TransportRolloverAction.generateRolloverIndexName("index"), equalTo("index-1"));
```
Maybe s/"index"/index/ and make all the names random? I guess you'd just have to make sure that `index` didn't end in a number but you could do that with `randomAsciiOfLength(10) + "A"` or something.

areek replied:
Thanks for the suggestion!

areek commented:
Thanks @nik9000  and @s1monw for the review! I addressed the comments and would love another round of feedback. The rest test is still not passing and I am currently investigating. 
Updated PR: 
 - simplified the request to only ask for the `alias` to be switched 
 - added flag for simulation (don't rollover even if conditions are met)
 - enhanced the response to include condition status, whether the index was rolled over and whether the rollover index was automatically created.
 - nuked max size condition

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -0,0 +1,259 @@
+                                false));
+                        return;
+                    }
+                    if (conditionStatus.stream().allMatch(Map.Entry::getValue)) {
```
I think the conditions should be ORed?

areek replied:
Now the conditions are `OR`ed

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/Condition.java:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        @Override
+        public boolean matches(TimeValue value) {
```
I think the condition should just take an object like:

```Java

static class Stats {
  final long numDoc;
  final TimeValue indexCreated;
}
```

and we can add whatever we want later?

areek replied:
Thanks for the suggestion!

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -0,0 +1,259 @@
+            new ActionListener<IndicesStatsResponse>() {
+                @Override
+                public void onResponse(IndicesStatsResponse indicesStatsResponse) {
+                    final IndexMetaData sourceIndex = metaData.index(sourceIndexName);
```
can you move the calcualtions out into static methods so we can unittest them? Look at how I wrote `MetaDataCreateIndexServiceTests`

areek replied:
Thanks for the pointer, moved calculations to static methods.

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -0,0 +1,259 @@
+
+                                @Override
+                                public void onFailure(Throwable t) {
+                                    if (t instanceof IndexAlreadyExistsException) {
```
I don't think we need that log lines here just report to the user

s1monw commented on rest-api-spec/src/main/resources/rest-api-spec/api/indices.rollover.json:
```diff
@@ -0,0 +1,30 @@
+    "methods": ["PUT", "POST"],
+    "url": {
+      "path": "/{alias}/_rollover",
+      "paths": ["/{alias}/_rollover", "/{alias}/_rollover/{target_alias}"],
```
this still has target_alias

areek replied:
oops, removed

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -0,0 +1,259 @@
+        return updateRequest;
+    }
+
+    static String generateRolloverIndexName(String sourceIndexName) {
```
I am not sure if I like this. I think we should be strict here and ensure that the index name matches with `.*(_\d)*^` and strip it off only at the end. If we hit `NFE` we should rethrow!

areek replied:
made this strict by checking if the index name matches `^.*-(\d)+$` first.

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    static CreateIndexClusterStateUpdateRequest prepareCreateIndexRequest(final String targetIndexName,
+                                                                          final RolloverRequest rolloverRequest) {
```
so we solely rely on the template the index matches right? I think we have to document this explicitly and also add the ability to pass an IndexRequest in the json body to override stuff in the template?

areek replied:
Should we just serialize the entire source to construct the create index request later, as we don't know the name of the target index until we resolve the name of the underlying index?

s1monw commented:
@areek it looks great. I left some minors, I think we also need to add body support for the create index request....

areek commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -0,0 +1,234 @@
+/**
+ * Main class to swap the index pointed to by an alias, given some predicates
+ */
+public class TransportRolloverAction extends TransportMasterNodeAction<RolloverRequest, RolloverResponse> {
```
@s1monw Maybe we should refactor out the condition checking to another Transport action (`TransportConditionalAction`)? such that an abstract class can execute the stats request and check the request conditions, while a subclass can implement the actions to execute upon satisfying the conditions?

s1monw replied:
I think it's fine to be in one action

areek commented:
Thanks @s1monw for the feedback! I added body support for the underlying create index request to override matching index template settings (though unlike create index request overriding custom index metadata is not supported in the body). Added documentation and updated the PR description. Would  be awesome to have this reviewed :)

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:
```diff
@@ -0,0 +1,195 @@
+        PARSER.declareField((parser, request, parseFieldMatcherSupplier) ->
+                Condition.PARSER.parse(parser, request.conditions, parseFieldMatcherSupplier),
+            new ParseField("conditions"), ObjectParser.ValueType.OBJECT);
+        PARSER.declareField((parser, request, parseFieldMatcherSupplier) ->
```
nice idea!!!

s1monw replied:
maybe we should do the same for `_shrink` as a followup.

areek replied:
Yeay :), opened https://github.com/elastic/elasticsearch/pull/18802

s1monw commented on docs/reference/indices/rollover-index.asciidoc:
```diff
@@ -0,0 +1,84 @@
+	"new_index": "index-2", <2>
+	"rolled_over": true, <3>
+	"simulated": false, <4>
+	"rollover_index_created": true, <5>
```
IMO if that index exists the rollover should fail!

areek replied:
nuked this option from the response, since now we fail if the rollover index existed before

s1monw commented on rest-api-spec/src/main/resources/rest-api-spec/api/indices.rollover.json:
```diff
@@ -0,0 +1,30 @@
+    "documentation": "http://www.elastic.co/guide/en/elasticsearch/reference/master/indices-rollover-index.html",
+    "methods": ["POST"],
+    "url": {
+      "path": "/{index}/_rollover",
```
I am still torn on if we should force the user into our naming pattern. Today we require the user to start with `${some_name}-${some_number}` and we increment that number on each rollover. I think we should make it optional for the user to specify the next indices name maybe via `/{index}/_rollover/{new_index}` and  then skip the pattern enforcement.  @clintongormley what do you think?

s1monw replied:
also we should name this `/{alias}/_rollover`

areek replied:
Now we allow users to specify the new index name in the rollover request. If not specified we require the old index to have `${some_name}-${some_number}` format.

areek replied:
I changed the endpoint to `/{alias}/_rollover` in the rest-api-spec

s1monw commented:
i think it looks great! @clintongormley can you take a look too please

clintongormley commented:
>  creating the index if it does not exist. 

I think it should fail if the index already exists.

> Using the http GET method for the API runs the rollover in simulated mode, where request conditions can be checked without performing the actual rollover. Setting simulate=true as a request parameter also runs the request in simulated mode.

This will break in Console (and maybe the clients, depending on settings).  Doing a GET request (without simulate=true) can get changed to a POST request because of the body, and end up actually executing the rollover.  I'd change this simulate mode to always require the flag (which I'd call `dry_run`, as we use this elsewhere).


> The _rollover API is similar to indices-create-index and accepts settings, mappings and aliases
to override the index create request for a non-existent rolled over index.

I'm having second thoughts about this.  If you make changes in the body, then these changes won't be reflected in the next rollover unless you either (a) repeat the changes in the next rollover request or (b) update the templates.  I'm wondering if we should make templates the only way to configure the rolled-over index.

> I am still torn on if we should force the user into our naming pattern. Today we require the user to start with ${some_name}-${some_number} and we increment that number on each rollover. I think we should make it optional for the user to specify the next indices name maybe via /{index}/_rollover/{new_index} and then skip the pattern enforcement. @clintongormley what do you think?

I agree, eg what if the user wants to convert `mylogs` into a rolling index, or even `mylogs-01` to `mylogs-00001`? They should be able to do so with `POST mylogs_alias/_rollover/mylogs-00001` 

If the target index name is not provided, then we should require that the index name matches our pattern.

s1monw commented:
> This will break in Console (and maybe the clients, depending on settings). Doing a GET request (without simulate=true) can get changed to a POST request because of the body, and end up actually executing the rollover. I'd change this simulate mode to always require the flag (which I'd call dry_run, as we use this elsewhere).

++

s1monw commented:
> I'm having second thoughts about this. If you make changes in the body, then these changes won't be reflected in the next rollover unless you either (a) repeat the changes in the next rollover request or (b) update the templates. I'm wondering if we should make templates the only way to configure the rolled-over index.

templates will apply always - I really want this stuff to be per request since relying on templates make things very hard if you wanna do something different based on a condition that matched. lets say you have a condition `age > 2d AND numDocs < 1M` you might want to cut the number of shards in a half. the next request is `age < 2d AND numDocs >1M`  you might wanna bump stuff up? (that reminds me that you might need some boolean logic here 💃 - I think we should add MUST | SHOULD operators to the conditions as well to be really useful? Followup might also add conditions like day of week and things like this? maybe I am  just confused?

clintongormley commented:
> templates will apply always - I really want this stuff to be per request since relying on templates make things very hard if you wanna do something different based on a condition that matched. lets say you have a condition age > 2d AND numDocs < 1M you might want to cut the number of shards in a half. the next request is age < 2d AND numDocs >1M you might wanna bump stuff up? (that reminds me that you might need some boolean logic here 💃 - I think we should add MUST | SHOULD operators to the conditions as well to be really useful? Followup might also add conditions like day of week and things like this? maybe I am just confused?

No, I think this is a good idea, but rather than creating a new syntax, we could just support scripting and pass size, age, etc, in as parameters to the script.  This would allow ultimate flexibility.  We should make this index-matcher a reusable component that can be applied to other index management APIs.

But as a follow up, not in this PR. 

+1 to keeping the create-index body in this PR.

s1monw commented:
> No, I think this is a good idea, but rather than creating a new syntax, we could just support scripting and pass size, age, etc, in as parameters to the script. This would allow ultimate flexibility. We should make this index-matcher a reusable component that can be applied to other index management APIs.

huge +1

areek commented:
> I think it should fail if the index already exists.

Now we fail when the rollover index already exists.

> This will break in Console (and maybe the clients, depending on settings). Doing a GET request (without simulate=true) can get changed to a POST request because of the body, and end up actually executing the rollover. I'd change this simulate mode to always require the flag (which I'd call dry_run, as we use this elsewhere).

aah, I removed the option of using `GET` request for the rollover api and renamed `simulate` option to `dry_run`. Now the user has to explicitly set `dry_run=true` in the request params to execute rollover in dry-run mode.

> I agree, eg what if the user wants to convert mylogs into a rolling index, or even mylogs-01 to mylogs-00001? They should be able to do so with POST mylogs_alias/_rollover/mylogs-00001
If the target index name is not provided, then we should require that the index name matches our pattern.

Now users have the option to provide the name for the new rolling index in the rollover request.
```
POST logs_alias/_rollover/mylogs-1
```
and when the rolling index name is not specified, we fall back to requiring the old index name to have `{prefix}-{num}` format to generate the new rolling index name.

> No, I think this is a good idea, but rather than creating a new syntax, we could just support scripting and pass size, age, etc, in as parameters to the script. This would allow ultimate flexibility. We should make this index-matcher a reusable component that can be applied to other index management APIs.

That sounds great! I was wondering how to generalize this so other index management APIs can work on indices that match conditions. Though, I think we can do this in a separate PR. 

>  think we should add MUST | SHOULD operators to the conditions as well to be really useful?

Should we do this in a separate PR?

Thanks @s1monw and @clintongormley for the feedback! I updated the PR and the description according to the discussions, WDYT?

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverResponse.java:
```diff
@@ -0,0 +1,144 @@
+        return builder;
+    }
+
+    static final class Fields {
```
I think this class is unnecessary, just make the constants part of the outer class

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -0,0 +1,224 @@
+            .masterNodeTimeout(createIndexRequest.masterNodeTimeout())
+            .settings(createIndexRequest.settings())
+            .aliases(createIndexRequest.aliases())
+            .customs(createIndexRequest.customs())
```
i thought customs are not supported?

s1monw commented on core/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java:
```diff
@@ -0,0 +1,56 @@
+    public RestRolloverIndexAction(Settings settings, RestController controller, Client client) {
+        super(settings, client);
+        controller.registerHandler(RestRequest.Method.POST, "/{index}/_rollover", this);
+        controller.registerHandler(RestRequest.Method.POST, "/{index}/_rollover/{new_index}", this);
```
s/index/alias?

s1monw replied:
I think it should be `"/{alias}/_rollover/{new_index}"`

areek replied:
@s1monw when I changed this to `{alias}` instead of `{index}`, the rest tests fail with a request validation failure for missing index alias. It seems the first variable is read always keyed as `index`. seems like I am missing something?

s1monw replied:
can you show me what you changed?

s1monw replied:
well ok I guess it's fine then?

areek replied:
@s1monw applying this patch:
```diff
diff --git a/core/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java b/core/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java
index 7eecaaa..77fff1c 100644
--- a/core/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java
+++ b/core/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java
@@ -37,14 +37,15 @@ public class RestRolloverIndexAction extends BaseRestHandler {
     @Inject
     public RestRolloverIndexAction(Settings settings, RestController controller, Client client) {
         super(settings, client);
-        controller.registerHandler(RestRequest.Method.POST, "/{index}/_rollover", this);
-        controller.registerHandler(RestRequest.Method.POST, "/{index}/_rollover/{new_index}", this);
+        controller.registerHandler(RestRequest.Method.POST, "/{alias}/_rollover", this);
+        controller.registerHandler(RestRequest.Method.POST, "/{alias}/_rollover/{new_index}", this);
     }
 
     @SuppressWarnings({"unchecked"})
     @Override
     public void handleRequest(final RestRequest request, final RestChannel channel, final Client client) {
-        RolloverRequest rolloverIndexRequest = new RolloverRequest(request.param("index"), request.param("new_index"));
+        logger.info("request params: " + request.params());
+        RolloverRequest rolloverIndexRequest = new RolloverRequest(request.param("alias"), request.param("new_index"));
         if (request.hasContent()) {
             rolloverIndexRequest.source(request.content());
         }
```

I get a rollover request validation exception, when running rest test `indices.rollover/10_basic/Rollover index via API`:
```
{
	"error": {
		"root_cause": [{
			"type": "action_request_validation_exception",
			"reason": "Validation Failed: 1: index alias is missing;"
		}],
		"type": "action_request_validation_exception",
		"reason": "Validation Failed: 1: index alias is missing;"
	},
	"status": 400
}
```
Investigating the elasticsearch log reveals the request params: `{index=logs_search}`, instead of `{alias=logs_search}` causing a request validation exception.

s1monw commented:
I left some minors, LGTM - if @clintongormley is happy with the docs I don't need another review round once the last comments are fixed

clintongormley commented on docs/reference/indices/rollover-index.asciidoc:
```diff
@@ -0,0 +1,84 @@
+[[indices-rollover-index]]
```
[[indices-rollover-index]]
    == Rollover Index
 
    The rollover index API rolls an alias over to a new index when the existing
    index is considered to be too large or too old. 

    The API accepts a single alias name and a list of `conditions`.  The alias
    must point to a single index only.  If the index satisfies the specified
    conditions then a new index is created and the alias is switched to point to
    the new alias.


    [source,js]
    --------------------------------------------------
    PUT /logs-0001 <1>
    {
      "aliases": {
        "logs_write": {} 
      }
    }

    POST logs_write/_rollover <2>
    {
      "conditions": {
        "max_age":   "7d",
        "max_docs":  1000
      }
    }
    --------------------------------------------------
    // CONSOLE
    <1> Creates an index called `logs-0001` with the alias `logs_write`.
    <2> If the index pointed to by `logs_write` was created 7 or more days ago, or
        contains 1,000 or more documents, then the `logs-0002` index is created
        and the `logs_write` alias is updated to point to `logs-0002`.

    The above request might return the following response:

    [source,js]
    --------------------------------------------------
    {
      "old_index": "logs-0001", 
      "new_index": "logs-0002", 
      "rolled_over": true, <1>
      "dry_run": false, <2>
      "conditions": { <3>
        "[max_age: 7d]": false,
        "[max_docs: 1000]": true
      }
    }
    --------------------------------------------------
     <1> Whether the index was rolled over.
     <2> Whether the rollover was dry run.
     <3> The result of each condition.

    [float]
    === Naming the new index

    If the name of the existing index ends with `-` and a number -- e.g.
    `logs-0001` -- then the name of the new index will follow the same pattern,
    just incrementing the number (`logs-0002`). 

    If the old name doesn't match this pattern then you must specify the name for
    the new index as follows:

    [source,js]
    --------------------------------------------------
    POST my_alias/_rollover/my_new_index_name
    {...}
    --------------------------------------------------

    [float]
    === Defining the new index

    The settings, mappings, and aliases for the new index will be taken from any
    matching <<indices-templates,index templates>>. Additionally, the body of the
    request accepts `settings`, `mappings`, and `aliases` just like the 
    <<indices-create-index,create index>> API, which will override any values
    set in matching index templates:

    [source,js]
    --------------------------------------------------
    POST /logs_writes/_rollover
    {
      "conditions" : {
        "max_age": "7d",
        "max_docs": 1000
      },
      "settings": { <1>
        "index.number_of_shards": 2 <1>
      }
    }
    --------------------------------------------------
    // CONSOLE
    <1> Set settings to override matching index template, `mappings` and `aliases` can also be provided.
     
     [float]
     === Dry run

    The rollover API supports `dry_run` mode, where request conditions can be
    checked without performing the actual rollover:

    [source,js]
    --------------------------------------------------
    POST /logs_writes/_rollover?dry_run
    {
      "conditions" : {
        "max_age": "7d",
        "max_docs": 1000
      }
    }
    --------------------------------------------------
    // CONSOLE

clintongormley commented:
I improved the docs - please check that they are actually correct :)

s1monw commented:
@clintongormley I don't see you comment about accepting the full create index body here? did I miss something?

clintongormley commented:
@s1monw it's here: https://github.com/elastic/elasticsearch/pull/18802#issuecomment-225667500

clintongormley commented:
And I've just realised my comment is in the wrong place.  It was meant to apply to this issue.  Copied here:

> @areek Sorry for the late comment but I think that the rollover API should accept a full create-index body, rather than be limited to settings and aliases, otherwise you have to use index templates regardless.

In fact I may have misunderstood completely and the above is already the case on the rollover API?

areek commented:
Thanks @clintongormley for the awesome documentation, it LGTM. I updated the docs 

> In fact I may have misunderstood completely and the above is already the case on the rollover API?

Yes, it is already the case for the rollover API and works as documented.
